content/articles/building-and-integrating-spring-interceptor-in-java/index.md

```diff
@@ -0,0 +1,123 @@
+
+For us to use Interceptors in Spring Boot, we need to add the web and Lombok dependencies in our `pom.xml` configuration file. The web dependency allows us to interact with services over HTTP while Lombok gives us access to the `@Slf4j` annotation which we will use later for logging.
+
+```xml
+<dependency>
+  <groupId>org.springframework.boot</groupId>
+  <artifactId>spring-boot-starter-web</artifactId>
+</dependency>
+<dependency>
+  <groupId>org.projectlombok</groupId>
+  <artifactId>lombok</artifactId>
+</dependency>
+```
+### Interceptor methods
+To demonstrate the operations of a Spring Interceptor, we will simulate a small request-response project. To begin, create a Spring Boot project. There are several articles on how to set up a Spring Boot project on the Section website. Do well to check for one or all of them.
+
+Now it is time to introduce the `Interceptor` methods. Spring Interceptors thrive on three methods of operation:
+- preHandle(): To carry out some actions on a request before sending it to the controller
+- postHandle(): To carry out actions on a response coming from the controller, before sending to requesting client
+- afterCompletion(): To carry out actions at the end of the request-response interaction
+
+In our project, all we will do is add log messages to the requests and responses to show how interception works.
+#### preHandle()
+The name describes what the method does: method invocation before a request is handled. It is at this point that we log information regarding the parameters of the request. We are logging this info using a simple Slf4j logger by simply annotating the class with the `@Slf4j` annotation:
+```java
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@Slf4j
+public class CustomInterceptor implements HandlerInterceptor {
+
+    @Override
+    public boolean preHandle(
+            HttpServletRequest req,
+            HttpServletResponse res) throws Exception {
+
+        log.info("[preHandle() method called during request handling {}", req);
+
+        return true;
+    }
+}
+
+```
+Now, let us look at the next method after `preHandle()`
+#### postHandle()
+This method is invoked by the `interceptor` after the request has been handled but just before the `DispatcherServlet` passes the response to the client. If we like, at this point, we could add more attributes to the response. We can also get to record the time of processing the request. Just like with the preHandle(), we will be logging a message in the method execution.
+```java
+// in the LoogingInterceptor class...
+@Override
+public void postHandle(
+  HttpServletRequest req,
+  HttpServletResponse res,
+  ModelAndView modelAndView) throws Exception {
+
+    log.info("postHandle() method called during response return {}", res);
+}
```

Comment:
This does not properly implement the `postHandle` method of the `HandlerInterceptor` interface. The interface requires the following method signature:

```java
void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler, ModelAndView modelAndView) throws Exception
```

Please fix this.